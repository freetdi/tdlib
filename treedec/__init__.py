from .graph import *
from .pp import *
from .greedy import *

from .greedy import fi

