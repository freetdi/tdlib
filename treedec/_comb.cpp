// Felix Salfelder, 2021
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//

// various algorithm combination wrappers

#include "boost_compat.h"
#include <boost/graph/graph_traits.hpp>
#include "graph_py.hpp"
#include "directed_view.hpp"
#include "induced_subgraph.hpp"
#include "elimination_orderings.hpp"
#include "marker.hpp"
#include "impl/greedy_heuristic.hpp"
#include <boost/python.hpp>
#include <boost/python/list.hpp>
#include <vector>
#include <set>
#include "graph_traits.hpp"
#include "boost_graph.hpp"
#include "algo_py.hpp"
#include "combinations.hpp"

using treedec::pending::PP_FI_TM;
using treedec::pending::PP_FI;
using treedec::algo::default_config;

typedef PP_FI<_balsvu, default_config> _ppfi_balsvu;
typedef PP_FI<_balsvd, default_config> _ppfi_balsvd;
typedef PP_FI<_balvvu, default_config> _ppfi_balvvu;
typedef PP_FI<_balvvd, default_config> _ppfi_balvvd;

typedef PP_FI_TM<_balsvu, default_config> _ppfitm_balsvu;
typedef PP_FI_TM<_balsvd, default_config> _ppfitm_balsvd;
typedef PP_FI_TM<_balvvu, default_config> _ppfitm_balvvu;
typedef PP_FI_TM<_balvvd, default_config> _ppfitm_balvvd;

BOOST_PYTHON_MODULE(_comb)
{ itested();
	namespace py = boost::python;

//	to_python_converter<std::pair<int, bool>, pair2tuple<int, bool> >(); // hack
//	to_python_converter<std::pair<int, int>, pair2tuple<int, int> >();
//	to_python_converter<std::pair<long, long>, pair2tuple<long, long> >();
//	to_python_converter<std::pair<unsigned, unsigned>, pair2tuple<unsigned, unsigned> >();
//	to_python_converter<std::pair<unsigned long, unsigned long>, pair2tuple<unsigned long, unsigned long> >();

	COMMON_ALGO_IFACE(_ppfitm, _balvvu);
	COMMON_ALGO_IFACE(_ppfitm, _balvvd);
	COMMON_ALGO_IFACE(_ppfitm, _balsvu);
	COMMON_ALGO_IFACE(_ppfitm, _balsvd);

	COMMON_ALGO_IFACE(_ppfi, _balvvu);
	COMMON_ALGO_IFACE(_ppfi, _balvvd);
	COMMON_ALGO_IFACE(_ppfi, _balsvu);
	COMMON_ALGO_IFACE(_ppfi, _balsvd);
}
