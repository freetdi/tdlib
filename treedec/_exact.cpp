// Felix Salfelder, 2021
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#include "boost_compat.h"

#include <boost/python.hpp>
#include <boost/python/list.hpp>
#include <vector>
#include <set>

#include "graph_py.hpp"
#include "exact_cutset.hpp"
#include "exact_ta.hpp"
#include "util_py.hpp"
#include "algo_py.hpp"

namespace py = boost::python;
using treedec::exact_ta;
using treedec::draft::exact_decomposition;
using treedec::algo::default_config;

typedef treedec::draft::exact_cutset<_balsvu, default_config> _cutset_balsvu;
typedef treedec::draft::exact_cutset<_balsvd, default_config> _cutset_balsvd;
typedef treedec::draft::exact_cutset<_balvvu, default_config> _cutset_balvvu;
typedef treedec::draft::exact_cutset<_balvvd, default_config> _cutset_balvvd;

template<class G>
struct cfg_8 : public default_config<G> {
	static constexpr unsigned max_vertex_index=255u;
};

typedef exact_ta<_balvvu, cfg_8> _ta_balvvu;
typedef exact_ta<_balsvu, cfg_8> _ta_balsvu;

typedef exact_decomposition<_balvvu, default_config, exact_ta> _ex17_balvvu;
typedef exact_decomposition<_balsvu, default_config, exact_ta> _ex17_balsvu;



BOOST_PYTHON_MODULE(_exact)
{ itested();

//	to_python_converter<std::pair<int, bool>, pair2tuple<int, bool> >(); // hack
//	to_python_converter<std::pair<int, int>, pair2tuple<int, int> >();
//	to_python_converter<std::pair<long, long>, pair2tuple<long, long> >();
//	to_python_converter<std::pair<unsigned, unsigned>, pair2tuple<unsigned, unsigned> >();
//	to_python_converter<std::pair<unsigned long, unsigned long>, pair2tuple<unsigned long, unsigned long> >();

	COMMON_ALGO_IFACE(_cutset, _balvvu);
	// COMMON_ALGO_IFACE(_cutset, _balvvd); does not compile.
	COMMON_ALGO_IFACE(_cutset, _balsvu);
	// COMMON_ALGO_IFACE(_cutset, _balsvd); does not compile.

	COMMON_ALGO_IFACE(_ta, _balvvu);
	COMMON_ALGO_IFACE(_ta, _balsvu);

	COMMON_ALGO_IFACE(_ex17, _balvvu);
	COMMON_ALGO_IFACE(_ex17, _balsvu);
}
