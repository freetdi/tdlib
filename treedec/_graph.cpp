// Felix Salfelder, 2017, 2021
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 2, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//

#include "boost_compat.h"

#include <set>
#include <typeinfo>
#include <vector>
#include <boost/python.hpp>
#include <boost/python/list.hpp>
#include "trace.hpp"
#include "graph_iter.hpp"
#include "graph_py.hpp"
#include "boost_graph.hpp"
#include "util_py.hpp"


// unneeded. probably
// typedef wrap::graph<graph_bal, boost::identity_property_map > graph_bal_wrap;
// typedef wrap::graph<graph_balu, boost::identity_property_map > graph_balu_wrap;

using namespace boost::python; // BUG
namespace py = boost::python;

typedef std::vector< boost::tuple< unsigned,
		  std::set<unsigned> > > pp_data_type;

template<class T1, class T2>
struct pair2tuple {
  static PyObject* convert(const std::pair<T1, T2>& pair) {
    return incref(py::make_tuple(pair.first, pair.second).ptr());
  }
};

template<class T>
struct addedgeret {
  static PyObject* convert( std::pair< std::pair<T, T>, bool> const& r) {
    auto p = py::make_tuple(r.first.first, r.first.second);

    return incref(py::make_tuple(p, r.second).ptr());
  }
};

#if 0
struct makeunsignedlong {
  static PyObject* convert(unsigned long const & x) {
    return incref(new PyObject()); // py::make_integer(x));
  }
};
struct makeunsigned {
  static PyObject* convert(unsigned const & x) {
    return incref(new PyObject()); // py::make_integer(x));
  }
};
struct makelong {
  static PyObject* convert(long const & x) {
    return incref(new PyObject()); // py::make_integer(x));
  }
};
#endif

template<class graph>
boost::python::object preprocessing(graph& g)
{
	 pp_data_type b;
    size_t lb_bs = g.preprocessing(b); // pass pythonstuff?
	 boost::python::list a, ret;

	 /// YUCK
	 // pass something resonable to PP
	 // WRONG PLACE. has nothing to do with python.
	 for(auto const& i: b){
		 boost::python::list c;
		 unsigned j = boost::get<0>(i);
		 c.append(g.maphack(j)); // this is the eliminated node.
		                         // bad/FIXME: the order is lost.

		 auto const& bagi=boost::get<1>(i);
		 for(auto const& j : bagi){
			 c.append(g.maphack(j));
		 }

		 a.append(c);
	 }

	 ret.append(g); // FIXME (remove)
	 ret.append(a);
	 ret.append(int(lb_bs)-1); // yuck
	 return ret;
}

// BUG? detail..
static PyObject *pythonExceptionType = NULL;
static PyObject* createExceptionClass(const char* name, PyObject* baseTypeObj = PyExc_Exception)
{
    using std::string;
    namespace bp = boost::python;

    const string scopeName = bp::extract<string>(bp::scope().attr("__name__"));
    const string qualifiedName0 = scopeName + "." + name;
    PyObject* typeObj = PyErr_NewException(qualifiedName0.c_str(), baseTypeObj, 0);
    if (!typeObj) bp::throw_error_already_set();
    bp::scope().attr(name) = bp::handle<>(bp::borrowed(typeObj));
    return typeObj;
}

namespace detail{

#if 0
unsigned stringtotype(const std::string& type)
{
	unsigned typenumber=graph::b_default;
	if(type==""){
	}else if(type=="badjl_vvd"){
		typenumber = graph::b_badjl_vvd;
	}else{ untested();
		throw exception_invalid("type");
	}
	return typenumber;
}

template<class what>
static boost::shared_ptr<what> graph_factory(
		boost::python::list const& x, boost::python::list const& edgl, std::string const& type)
{ untested();
	typedef boost::python::stl_input_iterator<typename what::vertex_descriptor> iterator_type;
	iterator_type i(x);
	iterator_type e;
	unsigned typenumber=stringtotype(type);

	typedef typename what::vertex_descriptor vd;
	typedef ::detail::pyedge_input_iterator<std::pair<vd, vd> > pi;

	trace1("fact", typenumber);
	return boost::shared_ptr<what>(new what(pi(edgl), pi(), i, e, typenumber));
}

struct below_checker{
	below_checker(unsigned max=-1u):_m(max){}
	bool operator()(unsigned what) const{return what<_m;}

	unsigned _m;
};

template<class what>
static boost::shared_ptr<what> graph_factory_n(
		 unsigned max, boost::python::list const& y, std::string const& type)
{
	trace1("fact_n", max);
	unsigned typenumber=stringtotype(type);

	typedef ::detail::pyedge_input_iterator<std::pair<int, int>, below_checker > pi;

	below_checker C(max);

	trace1("fact", typenumber);
	return boost::shared_ptr<what>(
			new what(pi(y,C), pi(), max, typenumber));
}

template<class what>
static boost::shared_ptr<what> graph_factory_n0(
		unsigned x, boost::python::list const& y)
{
	return graph_factory_n<what>(x, y, "");
}
#endif

template<class what>
static boost::shared_ptr<what> graph_factory0(
		boost::python::list const& x, boost::python::list const& y)
{
	return graph_factory<what>(x, y, "");
}

// PyObject* ex_invalid_pytype=NULL;
//
// void trans_ex(exception_invalid const &e)
// { itested();
// 	assert(ex_invalid_pytype!=NULL);
// 	boost::python::object pythonExceptionInstance(e);
// 	PyErr_SetObject(ex_invalid_pytype, pythonExceptionInstance.ptr());
// 	// PyErr_SetString(PyExc_RuntimeError, e.what());
// }

static void translate_exception_invalid(exception_invalid const &e)
{
    using namespace boost;
    python::object exc_t(python::handle<>(python::borrowed(pythonExceptionType)));
    exc_t.attr("cause") = python::object(e); // add the wrapped exception to the Python exception
//    exc_t.attr("what") = python::object(e.what()); // for convenience
    PyErr_SetString(pythonExceptionType, e.what()); // the string is used by print(exception) in python
}

template<class what>
static boost::shared_ptr<what> init_boost_graph(size_t n)
{ untested();
	return boost::shared_ptr<what>(new what(n));
}

template<class what>
static boost::shared_ptr<what> init_boost_graph0( boost::python::list const& E, size_t n)
{
//	typedef boost::python::stl_input_iterator<typename what::vertex_descriptor> iterator_type;
//	iterator_type i(x);
//	iterator_type e;

	typedef typename boost::graph_traits<what>::vertex_descriptor vd;
	typedef ::detail::pyedge_input_iterator<std::pair<vd, vd> > pi;

	what* ng = new what(pi(E), pi(), n);
	return boost::shared_ptr<what>(ng);
}


#if 0
// want: const reference.
// not possible in python, maybe
template<class P>
typename P::value_type get_first(P const& p){
	return p.first;
}
template<class P>
typename P::value_type get_second(P const& p){
	return p.second;
}
#endif

namespace py=boost::python;
py::object get_first(std::pair<py::object, py::object>& p)
{
	return p.first;
}
namespace py=boost::python;
py::object get_second(std::pair<py::object, py::object>& p)
{
	return p.second;
}

} // detail

#define COMMON_GRAPH_PY_IFACE(gid) \
		.def(init<size_t>()) \
		.def(init<std::string const&>()) \
		.def(init<size_t, std::string const&>()) \
		.def("backend_typename", & gid ::backend_typename) \
		.def("num_vertices",     & gid ::num_vertices) \
		.def("num_edges",        & gid ::num_edges) \
		.def("add_edge",         & gid ::add_edge) \
		.def("vertices",         & gid ::vertices) \
		.def("edges",            & gid ::edges) \
		; \
	typedef std::pair<gid ::vertex_iterator, gid ::vertex_iterator> gid ## _vip; \
	class_< gid ## _vip >( #gid "_vertex_range", no_init ) \
		 .def("__iter__" , py::range(& gid ## _vip::first, & gid ## _vip::second)) \
		; \
	typedef std::pair<gid ::edge_iterator, gid ::edge_iterator> gid ## _eip; \
	class_< gid ## _eip >( #gid "_edge_range", no_init ) \
		 .def("__iter__" , py::range(& gid ## _eip::first, & gid ## _eip::second)) \
		 ;
//	def("preprocessing", preprocessing<gid>)
//

template<class T, StringLiteral n>
static std::string make_string(T const& g)
{
//	return typeid(g).name();
	auto v = boost::num_vertices(g);
	auto e = boost::num_edges(g);
	return std::string(n.value) + " on " + std::to_string(v) + " vertices "
		+ "with " + std::to_string(e) + " edges";
}

#define COMMON_BOOST_GRAPH_PY_IFACE(gid) \
	class_< gid, boost::noncopyable >(#gid, init<>()) \
		.def("__init__", make_constructor(::detail::init_boost_graph< gid >)) \
		.def("__init__", make_constructor(::detail::init_boost_graph0< gid >)) \
		.def(init<size_t>()) \
		.def("num_vertices", &boost_num_vertices< gid >) \
		.def("num_edges",    &boost_num_edges< gid >) \
		.def("vertices",     &boost_vertices< gid >) \
		.def("edges",        &boost_edges< gid >) \
		.def("add_vertex",   &boost_add_vertex< gid >) \
		.def("add_edge",     &boost_add_edge< gid >) \
		.def("__repr__",     &make_string<gid, #gid> ) \
		; \
	typedef pair_<typename boost::graph_traits<gid>::vertex_iterator > gid ## _vip; \
	class_< gid ## _vip >( #gid "_vertex_range", no_init ) \
		 .def("__iter__" , py::range(& gid ## _vip::first, & gid ## _vip::second)) \
		; \
	typedef pair_<typename boost::graph_traits<gid>::edge_iterator > gid ## _eip; \
	class_< gid ## _eip >( #gid "_edge_range", no_init ) \
		 .def("__iter__" , py::range(& gid ## _eip::first, & gid ## _eip::second)) \
		 ;

#define COMMON_BOOST_GRAPH_PY_IFACE_WIP(gid) \
	class_< gid, boost::noncopyable >(#gid, init<>()) \
		.def("__init__", make_constructor(::detail::init_boost_graph< gid >)) \
		.def("__init__", make_constructor(::detail::init_boost_graph0< gid >)) \
		.def(init<size_t>()) \
		.def("num_vertices", &boost_num_vertices< gid >) \
		.def("num_edges",    &boost_num_edges< gid >) \
		.def("vertices",     &boost_vertices< gid >) \
		.def("edges",        &boost_edges< gid >) \
		.def("add_vertex",   &boost_add_vertex< gid >) \
		.def("add_edge",     &boost_add_edge< gid >) \
		.def("__repr__",     &make_string<gid, #gid> ) \
		; \
	typedef pair_<typename boost::graph_traits<gid>::vertex_iterator > gid ## _vip; \
	class_< gid ## _vip >( #gid "_vertex_range", no_init ) \
		 .def("__iter__" , py::range(& gid ## _vip::first, & gid ## _vip::second)) \
		;
	//typedef pair_<typename boost::graph_traits<gid>::edge_iterator > gid ## _eip; \
	//class_< gid ## _eip >( #gid "_edge_range", no_init ) \
	//	 .def("__iter__" , py::range(& gid ## _eip::first, & gid ## _eip::second)) \
	//	 ;

template<class G>
size_t boost_num_vertices(G const& g)
{
	return boost::num_vertices(g);
}
template<class G>
size_t boost_num_edges(G const& g)
{
	return boost::num_edges(g);
}



template<class G>
struct edg2pair{
	edg2pair(G const& g) : _g(g)
	{
	}
	typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
    typedef std::pair<vertex_descriptor, vertex_descriptor> edge_descriptor;
	template<class E>
	// std::pair<vertex_descriptor, vertex_descriptor>
	std::pair<unsigned, unsigned>
	operator()(E edg) const {
		auto s=boost::source(edg, _g);
		auto t=boost::target(edg, _g);
		return std::make_pair(s, t);
	}
	//idmap_type const& _m;
	G const& _g;
};


class my_vd{
};


template<class edge_descriptor>
using edge_iterator_ = any_iterator<const edge_descriptor,
			  std::input_iterator_tag,
			  const edge_descriptor>;

template<class G>
pair_< edge_iterator_< pair_<typename boost::graph_traits<G>::vertex_descriptor> > >
boost_edges(G const& g)
{
	typedef typename boost::graph_traits<G>::vertex_descriptor vertex_descriptor;
	typedef std::pair<vertex_descriptor, vertex_descriptor> edge_descriptor;
	typedef typename boost::graph_traits<G>::edge_iterator geit_t;

	typedef any_iterator<const edge_descriptor,
			  std::input_iterator_tag,
			  const edge_descriptor> edge_iterator;

	auto x = boost::edges(g);
	geit_t b(x.first);
	geit_t e(x.second);

	// translation
	auto bt=make_transform_iterator(b, edg2pair<G>(g));
	auto et=make_transform_iterator(e, edg2pair<G>(g));
	return std::make_pair(edge_iterator(bt), edge_iterator(et));
}


BOOST_PYTHON_MODULE(_graph)
{
	namespace py = boost::python;

	//py::to_python_converter<my_vd, makemyvd>();
#if 0
	py::to_python_converter<const unsigned, makeunsigned>();
	py::to_python_converter<unsigned, makeunsigned>();

	py::to_python_converter<const long, makelong>();

	py::to_python_converter<unsigned long, makeunsignedlong>();
	py::to_python_converter<const unsigned long, makeunsignedlong>();
	py::to_python_converter<long, makelong>();
#endif

#if 1 // TMP. add_edge return type
	py::to_python_converter<std::pair<int, bool>, pair2tuple<int, bool> >(); // hack
#endif
	py::to_python_converter<std::pair<int, int>, pair2tuple<int, int> >();
	py::to_python_converter<std::pair<long, long>, pair2tuple<long, long> >();
	py::to_python_converter<std::pair<unsigned, unsigned>, pair2tuple<unsigned, unsigned> >();
	py::to_python_converter<std::pair<unsigned long, unsigned long>, pair2tuple<unsigned long, unsigned long> >();

	py::to_python_converter<std::pair<std::pair< unsigned long, unsigned long >, bool>, addedgeret<unsigned long> >();

// 	std::pair<boost::detail::undirected_edge_iter<std::_List_iterator<boost::list_edge<unsigned long, boost::no_property> >

//    to_python_converter<Vertex, cached_vertex_object<Graph> >();
//    to_python_converter<Edge, cached_edge_object<Edge> >();

	class_<std::pair<py::object, py::object> >("edge")
		    .def("source", &::detail::get_first)
			 .def("target", &::detail::get_second)
			 ;

//	class_<graph_bal_wrap>("graph_bal_wrap", init<>())
//		//.def(init< boost::python::list const&, boost::python::list const&>())
//		.def("__init__", make_constructor(::detail::graph_factory_n<graph_bal_wrap>))
//		.def("__init__", make_constructor(::detail::graph_factory_n0<graph_bal_wrap>))
//		COMMON_GRAPH_PY_IFACE(graph_bal_wrap)
//		;

//	class_<graph_balu_wrap>("graph_balu_wrap", init<>())
//		//.def(init< boost::python::list const&, boost::python::list const&>())
//		.def("__init__", make_constructor(::detail::graph_factory_n<graph_balu_wrap>))
//		.def("__init__", make_constructor(::detail::graph_factory_n0<graph_balu_wrap>))
//		COMMON_GRAPH_PY_IFACE(graph_balu_wrap)
//		;
//
	typedef std::pair<wrap::graph<_balvvu>::edge_iterator,
	                  wrap::graph<_balvvu>::edge_iterator> graph_eip_1;
	class_< graph_eip_1 >( "graph_edge_range", no_init ) \
		 .def("__iter__" , range(& graph_eip_1::first, & graph_eip_1::second)) \
		; \

	typedef std::pair<wrap::graph<_balvvu>::raw_vertex_iterator,
	                  wrap::graph<_balvvu>::raw_vertex_iterator> graph_vip_1;
	class_< graph_vip_1 >( "graph_vertex_range", no_init ) \
		 .def("__iter__" , range(& graph_vip_1::first, & graph_vip_1::second)) \
		;

#if 0 // ?

	typedef std::pair<graph ::raw_vertex_iterator, graph ::raw_vertex_iterator> graph_vip; \
	class_< graph_vip >( "graph_vertex_range", no_init ) \
		 .def("__iter__" , range(& graph_vip::first, & graph_vip::second)) \
		; \
	typedef std::pair<graph ::edge_iterator, graph ::edge_iterator> graph_eip; \
	class_< graph_eip >( "graph_edge_range", no_init ) \
		 .def("__iter__" , range(& graph_eip::first, & graph_eip::second)) \
		 ;
#endif

// class_<graph>("graph", init<>())
// 	//.def(init< boost::python::list const&, boost::python::list const&>())
// 	.def("__init__", make_constructor(::detail::graph_factory_n<graph>))
// 	.def("__init__", make_constructor(::detail::graph_factory_n0<graph>))
//	   COMMON_GRAPH_PY_IFACE(graph)
// 	;
// 
// 	class_<subgraph>("subgraph", init<>())
// 		//.def(init< boost::python::list const&, boost::python::list const&>())
// 		.def("__init__", make_constructor(::detail::graph_factory0<subgraph>))
// 		.def("__init__", make_constructor(::detail::graph_factory<subgraph>))
// 		COMMON_GRAPH_PY_IFACE(subgraph)
// 		;
// 
// 	class_<Graph>("Graph", init<>())
// 		//.def(init< boost::python::list const&, boost::python::list const&>())
// 		.def("__init__", make_constructor(::detail::graph_factory0<Graph>))
// 		.def("__init__", make_constructor(::detail::graph_factory<Graph>))
// 		COMMON_GRAPH_PY_IFACE(Graph)
// 		;

#ifdef HAVE_GALA_GRAPH_H
	COMMON_BOOST_GRAPH_PY_IFACE_WIP(_gsgvvu16)
	COMMON_BOOST_GRAPH_PY_IFACE_WIP(_gsgvvu32)
	COMMON_BOOST_GRAPH_PY_IFACE_WIP(_gsgvvu64)
#endif

	COMMON_BOOST_GRAPH_PY_IFACE(_balvvd)
	COMMON_BOOST_GRAPH_PY_IFACE(_balsvd)
	COMMON_BOOST_GRAPH_PY_IFACE(_balvvu)
	COMMON_BOOST_GRAPH_PY_IFACE(_balsvu)



	class_<exception_invalid> cpp_exception_invalid_class("exception_invalid", init<std::string>());
	cpp_exception_invalid_class.def("what", &exception_invalid::what);

	pythonExceptionType = createExceptionClass("exception_invalid");
	boost::python::register_exception_translator<exception_invalid>(&::detail::translate_exception_invalid);
}

//template class wrap::graph<graph_bal, boost::identity_property_map>;
//template class wrap::graph<graph_balu, boost::identity_property_map>;
//template class tf::graph<boost::identity_property_map>;
