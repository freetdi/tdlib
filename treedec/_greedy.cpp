// Felix Salfelder, 2021
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the
// Free Software Foundation; either version 3, or (at your option) any
// later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//

// various greedy algorithm wrappers

#include "boost_compat.h"
#include <boost/graph/graph_traits.hpp>
#include "graph_py.hpp"
#include "directed_view.hpp"
#include "induced_subgraph.hpp"
#include "elimination_orderings.hpp"
// treedec::impl::bmdo<sg_dvv16> A(g16, _elimord);

#include "marker.hpp"
#include "impl/fill_in.hpp"
#include <boost/python.hpp>
#include <boost/python/list.hpp>
#include <vector>
#include <set>
#include "graph_traits.hpp"
#include "boost_graph.hpp"
#include "algo_py.hpp"

using treedec::impl::fillIn;
using treedec::algo::default_config;
using treedec::impl::bmdo;

typedef fillIn<_balsvu, default_config> _fi_balsvu;
typedef fillIn<_balsvd, default_config> _fi_balsvd;
typedef fillIn<_balvvu, default_config> _fi_balvvu;
typedef fillIn<_balvvd, default_config> _fi_balvvd;

typedef treedec::impl::bmdo<_balvvu> _bmd_balvvu;
typedef treedec::impl::bmdo<_balvvd> _bmd_balvvd;
typedef treedec::impl::bmdo<_balsvu> _bmd_balsvu;
typedef treedec::impl::bmdo<_balsvd> _bmd_balsvd;

#ifdef HAVE_GALA_GRAPH_H
typedef fillIn<_gsgvvu16, default_config> _fi_gsgvvu16;
typedef fillIn<_gsgvvu32, default_config> _fi_gsgvvu32;
typedef fillIn<_gsgvvu64, default_config> _fi_gsgvvu64;

typedef bmdo<_gsgvvu16> _bmd_gsgvvu16;
typedef bmdo<_gsgvvu32> _bmd_gsgvvu32;
typedef bmdo<_gsgvvu64> _bmd_gsgvvu64;
#endif

BOOST_PYTHON_MODULE(_greedy)
{ itested();
	namespace py = boost::python;

	COMMON_ALGO_IFACE(_fi, _balvvu);
	COMMON_ALGO_IFACE(_fi, _balvvd);
	COMMON_ALGO_IFACE(_fi, _balsvu);
	COMMON_ALGO_IFACE(_fi, _balsvd);

	COMMON_ALGO_IFACE(_bmd, _balvvu)
#ifdef HAVE_GALA_GRAPH_H
		    defgt(_bmd, _balvvu, _gsgvvu16)
		    defgt(_bmd, _balvvu, _gsgvvu32)
		    defgt(_bmd, _balvvu, _gsgvvu64)
#endif
	;
	COMMON_ALGO_IFACE(_bmd, _balvvd);
	COMMON_ALGO_IFACE(_bmd, _balsvu);
	COMMON_ALGO_IFACE(_bmd, _balsvd);

#ifdef HAVE_GALA_GRAPH_H
	COMMON_ALGO_IFACE(_fi, _gsgvvu16)
		    .def("get_treedec", &_fi_gsgvvu16::get_tree_decomposition<_balvvu_treedec>);
	COMMON_ALGO_IFACE(_fi, _gsgvvu32)
		    .def("get_treedec", &_fi_gsgvvu32::get_tree_decomposition<_balvvu_treedec>);
	COMMON_ALGO_IFACE(_fi, _gsgvvu64)
		    .def("get_treedec", &_fi_gsgvvu64::get_tree_decomposition<_balvvu_treedec>);

	COMMON_ALGO_IFACE(_bmd, _gsgvvu16)
		defgt(_bmd, _gsgvvu16, _balvvu)
		defgt(_bmd, _gsgvvu16, _gsgvvu32)
		defgt(_bmd, _gsgvvu16, _gsgvvu64);
	COMMON_ALGO_IFACE(_bmd, _gsgvvu32)
		defgt(_bmd, _gsgvvu32, _balvvu)
		defgt(_bmd, _gsgvvu32, _gsgvvu16)
		defgt(_bmd, _gsgvvu32, _gsgvvu64);
	COMMON_ALGO_IFACE(_bmd, _gsgvvu64)
		defgt(_bmd, _gsgvvu64, _balvvu)
		defgt(_bmd, _gsgvvu64, _gsgvvu16)
		defgt(_bmd, _gsgvvu64, _gsgvvu32);

#endif
}
