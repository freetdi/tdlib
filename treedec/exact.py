# (c) 2021 Felix Salfelder
# GPLv3+

from . import _graph as G
from . import _exact as ex

def cutset(g):
	if isinstance(g, G._balvvu):
		return ex._cutset_balvvu(g)
	elif isinstance(g, G._balsvu):
		return ex._cutset_balsvu(g)
#	elif isinstance(g, G._balvvd):
#		return ex._cutset_balvvd(g)
#	elif isinstance(g, G._balsvd):
#		return ex._cutset_balsvd(g)
	else:
		raise ValueError("cutset: can't handle " + str(g))

def ta(g):
	if isinstance(g, G._balvvu):
		return ex._ta_balvvu(g)
	elif isinstance(g, G._balsvu):
		return ex._ta_balsvu(g)
#	elif isinstance(g, G._balvvd):
#		return ex._ta_balvvd(g)
#	elif isinstance(g, G._balsvd):
#		return ex._ta_balsvd(g)
	else:
		raise ValueError("ta: can't handle " + str(g))

def ex17(g):
	if isinstance(g, G._balvvu):
		return ex._ex17_balvvu(g)
	elif isinstance(g, G._balsvu):
		return ex._ex17_balsvu(g)
#	elif isinstance(g, G._balvvd):
#		return ex._ex17_balvvd(g)
#	elif isinstance(g, G._balsvd):
#		return ex._ex17_balsvd(g)
	else:
		raise ValueError("ex17: can't handle " + str(g))
