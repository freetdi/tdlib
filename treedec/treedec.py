# (c) 2021 Felix Salfelder
# GPLv3+

from . import _treedec
