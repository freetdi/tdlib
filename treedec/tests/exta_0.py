
from treedec import _graph as gr
from treedec import _treedec as td
from treedec.exact import *

EU = [(0,1),(1,2),(2,3),(3,0)]

g = gr._balsvu(EU, 4)
print(g)
a = ta(g)
a.do_it()
print(a, a.bagsize())
t = td._balsvu_treedec()
a.get_treedec(t)

j=0
for i in t.vertices():
	print(i)
	B = t.get_bag(j);

	for k in B:
		print("", k, end="")
	print()
	j += 1

g = gr._balvvu(EU, 4)
print(g)
a = ta(g)
a.do_it()
t = td._balvvu_treedec()
a.get_treedec(t)
print(a, "bagsize", a.bagsize(), t)

for i in t.edges():
	print(i)

j=0
for i in t.vertices():
	print(i)
	B = t.get_bag(j);

	for k in B:
		print("", k, end="")
	print()
	j += 1
