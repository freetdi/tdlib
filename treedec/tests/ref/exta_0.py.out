_balsvu on 4 vertices with 4 edges
_ta_balsvu 3
0
 1 2 3
1
 0 1 3
_balvvu on 4 vertices with 4 edges
_ta_balvvu bagsize 3 _balvvu_treedec with 2 bags
(0, 1)
0
 1 2 3
1
 0 1 3
